contracts/evm/contracts/Oracle.sol

```diff
@@ -0,0 +1,258 @@
+
+        require(
+            sourcesThreshold <= numSourcesAggregated,
+            "Does not meet threshold for aggreagated sources"
+        );
+
+        if (checkpointIndex[pairId] > 0) {
+            Checkpoint memory currentCheckpoint = checkpoints[pairId][
+                checkpointIndex[pairId] - 1
+            ];
+            require(
+                currentCheckpoint.timestamp < lastUpdatedTimestamp,
+                "stale"
+            );
+        }
+        Checkpoint memory newCheckpoint = Checkpoint(
+            lastUpdatedTimestamp,
+            value,
+            aggregationMode,
+            numSourcesAggregated
+        );
+
+        checkpointIndex[pairId]++;
+        checkpoints[pairId].push(newCheckpoint);
+
+        return ();
+    }
+
+    function publishSpotEntries(SpotEntry[] calldata spotEntries) public {
+        for (uint256 i = 0; i < spotEntries.length; i++) {
+            publishSpotEntry(spotEntries[i]);
+        }
+    }
+
+    function getSpot(
+        bytes32 pairId,
+        AggregationMode aggregationMode,
+        bytes32[] memory sources
+    )
+        public
+        view
+        returns (
+            uint256 price,
+            uint256 decimals,
+            uint256 lastUpdatedTimestamp,
+            uint256 numSourcesAggregated
+        )
+    {
+        (SpotEntry[] memory entries, uint256 length, ) = getSpotEntries(
+            pairId,
+            sources
+        );
+        if (entries.length == 0) {
+            return (0, 0, 0, 0);
+        }
+        uint256 price = _aggregateSpotEntries(entries);
+        uint256 decimals = _getSpotDecimals(pairId);
+        uint256 lastUpdatedTimestamp = _maxSpotTimestamp(entries);
+        return (price, decimals, lastUpdatedTimestamp, length);
+    }
+
+    function getSpotEntries(bytes32 pairId, bytes32[] memory sources)
+        public
+        view
+        returns (
+            SpotEntry[] memory,
+            uint256,
+            uint256
+        )
+    {
+        uint256 lastUpdatedTimestamp = _getLatestSpotEntryTimestamp(
+            pairId,
+            sources
+        );
+        (SpotEntry[] memory entries, uint256 length) = _getAllSpotEntries(
+            pairId,
+            sources,
+            lastUpdatedTimestamp
+        );
+        return (entries, length, lastUpdatedTimestamp);
+    }
+
+    function _getSpotDecimals(bytes32 pairId) internal view returns (uint256) {
+        bytes32 baseCurrencyid = pairs[pairId].baseCurrencyId;
+        return currencies[baseCurrencyid].decimals;
+    }
+
+    function _maxSpotTimestamp(SpotEntry[] memory entries)
+        internal
+        pure
+        returns (uint256 latestTimestamp)
+    {
+        for (uint256 i = 0; i < entries.length; i++) {
+            if (entries[i].base.timestamp > latestTimestamp) {
+                latestTimestamp = entries[i].base.timestamp;
+            }
+        }
+        return latestTimestamp;
+    }
+
+    function _getLatestSpotEntryTimestamp(
```

Comment:
Remove and use getEntries + maxTimestamp; otherwise we have weird things e.g. with timestamp based filtering